### PR TITLE
[DistillationTrainer refactor] Switch the loss to GRPO's keys

### PR DESCRIPTION
**Pull request by qgallouedec**
*Item 21 (Group B) — stacked on PR 20. Part of #6449.*

The "switch" step of the layout migration: the loss now reconstructs from `prompt_ids`/`prompt_mask`/`completion_ids`/`completion_mask` instead of reading `input_ids`/`labels`/`prompts`/`prompt_attention_mask`.

- `compute_loss` and the Liger path build `input_ids = cat(prompt_ids, completion_ids)`, `attention_mask = cat(prompt_mask, completion_mask)`, slice logits by `logits_to_keep = completion_ids.size(1)`, and mask with the region-shaped `completion_mask`. `_get_teacher_logits` now takes `(input_ids, attention_mask)`.
- `completion_mask` is now emitted **region-shaped** `(B, C)` (was full-sequence at item 18), and `long` so it cats with `prompt_mask`. `_fill_buffer` counts `num_items` from `completion_mask`.
- Loss-value-preserving: `cat(prompt_ids, completion_ids) == input_ids`, and the same completion tokens are masked. The old keys stay *emitted* (the server + metrics still use them; they're removed with the buffer at item 29).
- Tests: `test_generated_batch_emits_completion_mask` now checks region shape + `cat(prompt_mask, completion_mask) == attention_mask`; `test_loss_normalizes` builds the GRPO key layout.

Verified: 53 tests pass (distillation + server); ruff clean.

---

> [!NOTE]
> **Medium Risk**
> Refactors the core distillation loss and Liger paths; behavior is intended to stay equivalent and is covered by updated tests, but any mistake in masking or logits slicing would change training objectives.
> 
> **Overview**
> **DistillationTrainer** now drives the JSD loss from GRPO-style tensors (`prompt_ids` / `prompt_mask` / `completion_ids` / `completion_mask`) instead of full-sequence `input_ids`, `labels`, and prompt-length slicing.
> 
> `compute_loss` and the Liger path **rebuild** `input_ids` and `attention_mask` via `cat` on those keys, slice logits with `logits_to_keep = completion_ids.size(1)`, and reduce with the **region-shaped** `completion_mask` `(B, C)`. `_get_teacher_logits` takes `(input_ids, attention_mask)` only. After generation, `_build_sequence_batch` emits `completion_mask` aligned with `completion_ids`; `_fill_buffer` sets the gradient-accumulation denominator from `completion_mask.sum()`. The collator sets `completion_mask` on the completion slice (not `labels != -100`).
> 
> Tests assert the new mask shape, `cat(prompt_mask, completion_mask) == attention_mask`, and GRPO-key batches for `num_items_in_batch` normalization.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit aea1399d96a15b7ccdfe3fd631779c97edc73a83. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/distillation/distillation_trainer.py [hunk 1]
```diff
@@ -983,18 +981,19 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
             loss = self._compute_liger_loss(model, inputs, num_items_in_batch=num_items_in_batch)
             return (loss, None) if return_outputs else loss
 
+        prompt_ids, prompt_mask = inputs["prompt_ids"], inputs["prompt_mask"]
+        completion_ids, completion_mask = inputs["completion_ids"], inputs["completion_mask"]
+        input_ids = torch.cat([prompt_ids, completion_ids], dim=1)
+        attention_mask = torch.cat([prompt_mask, completion_mask], dim=1)
+        logits_to_keep = completion_ids.size(1)  # only the completion tokens are trained on
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/distillation/distillation_trainer.py:988]: Intentional staging: the old keys stay emitted so the server still works, and the next PR (21b) freezes ServerDistillationTrainer self-contained before items 22/26/31 delete those base internals (full re-point at 58); its loss is a distinct external-teacher/sparse path, not a mechanical key-swap.

**Review by cmpatino — APPROVED** [review]
